Adicionado método para processar novas transações

GCastilho opened the pull request:
Listener:
-- O listener da currencyApi foi refatorado e agora é um módulo interno do self (resolves #19)
-- Implementado parser de requests POST no listener
-- Separado funções da currencyApi em get e post, não é mais possível acessar uma função projetada para GET usando POST e vice-versa
-- Adicionada a possibilidade de retornar um erro personalizado de um request em POST

Database:
-- Criado model para uma transaction
-- O sub-schema currencies agora tem o array de accounts em [currency].accounts
-- Adaptado código para a modificação no sucb-schema da currencies

currencyModule/new_transaction
-- Implementada função 'new_transaction' na currencyModule, para salvar transações recebidas em [currency].received
-- Ao concluir, é emitido um evento público 'new_transaction' no eventEmitter da currency
-- new_transaction retorna 404 se nenhum usuário foi encontrado para aquela account

currencyApi:
-- Implementado padrão init/constructor na currencyApi
-- Implementado EventEmitter interno na currencyApi em _events
-- Implementado EventEmitter público na currencyApi em events
-- Adicionado método 'init' new_transaction na currencyApi, que monitora os eventEmitters públicos dos módulos individuais pelo evento 'new_transaction' os retransmite no EventEmitter público da currencyApi, com o primeiro argumento sendo o nome da currency em questão

Bugfix:
-- Corrigido um bug no event handler da create_account

GCastilho committed:
Adicionado comando 'new_transaction' na currencyModule

-- Implementado padrão init/constructor na currencyApi
-- Implementado EventEmitter interno na currencyApi em _events
-- Implementado parser de requests POST no listener
-- Adicionado parser de 'new_transaction' no listener, que emite um 'new_transaction' no _event da currencyApi
-- Adicionado listener de 'new_transaction' na _events, que chama o comando 'new_transaction' da currency correspondente
-- Adicionado comando 'new_transaction' na currencyModule (ainda não implementado)

GCastilho committed:
Listener agora é um módulo do self

-- O listener da currencyApi foi refatorado e agora é um módulo interno do self
-- Separado funções da currencyApi em get e post, não é mais possível acessar uma função projetada para GET usando POST e vice-versa
-- Como o listener é um módulo da currencyApi, a função 'new_transaction' agora pode utilizar os métodos privados e diretamente chamar a new_transaction da currency correspondente, sem precisar emitir um evento para fazer essa ponte
-- This commit resolves #19

GCastilho committed:
Implementada função para salvar transações recebidas

Bugfix:
-- Corrigido um bug no event handler da create_account

Database:
-- Criado model para uma transaction
-- O sub-chema currencies agora tem o array de accoutns em [currency].accounts
-- Adaptado código para a modificação no sucb-chema da currencies

Create_accounts
-- Implementadad função 'new_transaction', para salvar transações recebidas em [currency].received

GCastilho committed:
new_transaction agora usa findOneAndUpdate

-- new_transaction agora usa findOneAndUpdate, para melhorar performance e impossibilitar race condition
-- Criado eventEmitter público para a currencyModule
-- new_transaction agora emite um evento público 'new_transaction' após receber e salvar a nova transação
-- Adicionada a possibilidade de retornar um erro de um request post
-- new_transaction agora retorna 404 se nenhum usuário foi encontrado para aquela account
-- Alguns bugfixes

GCastilho committed:
Evento de new_transaction agora é emitido do pela currencyApi

-- A currencyApi agora tem um eventEmitter público em this.events
-- O módulo new_transction da currencyApi monitora os eventEmitters públicos dos módulos individuais e os retransmite em seu eventEmitter, com o primeiro argumento sendo o nome da currency em questão

GCastilho committed:
Minor fixes

-- Removido console.log que printa uma msg ao receber uma nova transação
-- Minor fixes

gabr1gus commented on src/currencyApi/index.js:
```diff
@@ -40,10 +40,46 @@ class CurrencyApi {
+		/**
+		 * Instancia o EventEmitter público da currencyApi
+		 */
+		this.events = new this.events
```
eu nao tenho certeza se diferenciar o event emitter interno e publico por um " _ " é uma boa ideia, mas nao tenho uma sugestao melhor e nao acho que isso importa muito

GCastilho replied:
Isso é convenção em JS, métodos privados tem um "_" antes, já que não exit(iam) métodos privados de fato até recentemente

gabr1gus commented on src/currencyApi/listener/app.js:
```diff
@@ -1,48 +0,0 @@
-	app.listen(port, () => {
-		console.log(`CurrencyApi listener is up on port ${port}`)
-	})
-}
```
por que esse arquivo foi deletado ? foi substituido ?

gabr1gus replied:
entendi, o arquivo so foi movido e depois alterado para o que esta na pasta self

gabr1gus commented on src/currencyApi/self/listener/app.js:
```diff
@@ -0,0 +1,63 @@
+app.use(bodyParser.json({ extended: true }))
+
+module.exports = function init_listener_app() {
+	app.get('/:command/:currency', listener_get = (req, res) => {
```
nao sabia que se podia usar as rotas do express dessa forma, muito útil saber disso

GCastilho replied:
É o sistema de parâmetros de uma query, mto útil, pq da pra extrair o :\<coisa> como uma variável

gabr1gus commented on src/currencyApi/self/listener/functions/get/account_list.js:
```diff
@@ -0,0 +1,23 @@
+const Person = require('../../../../../db/models/person')
+
+module.exports = function get_account_list(currency, res) {
+	Person.find()
```
eu tenho quase certeza que esse find retorna mais dados do que vc quer , como object id e account de outra currency, recomendo o uso de projections, no mongodb shell seria algo como 
find( {},    //sim esse {} é necessario nesse caso
{_id: 0,   // especifica que o id nao vai vir, ela é inutil para essa operaçao
currencies : { currency : 1 }  //diz que somente a moeda currency dentro de currency ira vir
})
esse codigo nao esta formatado por que eu fiz de cabeça e pode estar errado, mas a ideia geral é especificar os campos desejados no find

GCastilho replied:
E qual o benefício de fazer desse modo? Pq não é como se o volume de dados fosse gigante

gabr1gus commented on src/currencyApi/self/listener/functions/get/index.js:
```diff
@@ -0,0 +1,19 @@
+require('fs').readdirSync(normalizedPath)
+	.forEach(filename =>
+		filename !== 'index.js' &&
+	(get_functions[filename.replace('.js', '')] = require(`./${filename}`))
```
eu acho incrivel a possibilidade de gerar funcoes dinamicamente baseado nos nomes de arquivos

GCastilho replied:
Na prática é uma propriedade, mas sim. Permite fazer um objeto em que cada método tá escrito em um arquivo, isso é excelente

(cortesia do JotaPê)

gabr1gus reviewed (approved):
o código esta bom, somente solicito que pense em mudar o find para um em que retorne apenas o necessário

GCastilho committed:
Regatorado get_account_list

-- Refatorado get_account_list para retornar apenas os campos necessários do documento
-- Também melhorada a organizaçao visual e indentação do modulo
